src/cli/commands/test/mod.rs

```diff
@@ -112,22 +151,120 @@ fn list_cairo_files(root: &PathBuf) -> Result<Vec<PathBuf>, String> {
 	.map(|cmd_output: ListOutput| cmd_output.files)
 }
 
-fn compile_and_list_entrypoints(path_to_code: PathBuf) -> Option<(PathBuf, PathBuf, Vec<String>)> {
-	match compile(&path_to_code) {
-		Ok(path_to_compiled) => match list_test_entrypoints(&path_to_compiled) {
-			Ok(entrypoints) => Some((path_to_code, path_to_compiled, entrypoints)),
-			Err(e) => {
-				eprintln!(
-					"Failed to list test entrypoints for file {}: {}",
-					path_to_compiled.display(),
-					e
-				);
-				None
+fn read_json_file(path: &PathBuf) -> Result<CacheJson, String> {
+	// Open the file in read-only mode with buffer.
+	let file = File::open(path).map_err(|op| format!("file does not exists {}", op))?;
+	let reader = BufReader::new(file);
+	let data =
+		serde_json::from_reader(reader).map_err(|op| format!("file does not exists {}", op))?;
+	return Ok(data);
+}
+
+fn compile_and_list_entrypoints(
+	cache: Result<CompiledCacheFile, String>,
+) -> Option<(PathBuf, PathBuf, Vec<String>)> {
+	match cache {
+		Ok(cache) => match cache.status {
+			CacheStatus::Cached => {
+				println!("Using cached compiled file");
+				let compiled_path = cache.path.clone();
+				let entrypoints =
+					list_test_entrypoints(&cache.path).expect("Failed to list entrypoints");
+				return Some((cache.path, compiled_path, entrypoints));
+			},
+			CacheStatus::Uncached => {
+				let compiled_path = compile(&cache.path).expect("Failed to compile");
+				let entrypoints =
+					list_test_entrypoints(&compiled_path).expect("Failed to list entrypoints");
+				return Some((cache.path, compiled_path, entrypoints));
 			},
 		},
-		Err(e) => {
-			eprintln!("{}", e);
-			None
+		Err(err) => {
+			eprintln!("{}", err);
+			return None;
+		},
+	}
+}
+
+fn create_compiled_contract_path(path_to_code: &PathBuf) -> PathBuf {
+	let filename = path_to_code.file_stem().expect("File does not have a file stem");
+
+	let cache_dir = dirs::cache_dir().expect("Could not make cache directory");
+	let mut path_to_compiled = PathBuf::new();
+	path_to_compiled.push(&cache_dir);
+	path_to_compiled.push("compiled-cairo-files");
+	path_to_compiled.push(filename);
+	path_to_compiled.set_extension("json");
+	return path_to_compiled;
+}
+
+fn dump_json_file(path: &PathBuf, data: &CacheJson) -> Result<(), String> {
+	let file = File::create(path).map_err(|op| format!("file does not exists {}", op))?;
+	let writer = BufWriter::new(file);
+	serde_json::to_writer_pretty(writer, data)
+		.map_err(|op| format!("file does not exists {}", op))?;
+	return Ok(());
```

Comment:
Moreover, you are only writing once there. So there is no actual need to use a buffer, you can just write to the file directly. It will do a single I/O operation.